File: LetsPlant/app/src/main/java/com/e/letsplant/activities/SignActivity.java
package com.e.letsplant.activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.lifecycle.ViewModelProvider;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.method.HideReturnsTransformationMethod;
import android.text.method.PasswordTransformationMethod;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.e.letsplant.R;
import com.e.letsplant.data.User;
import com.e.letsplant.data.UserViewModel;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;


import java.util.Objects;

public class SignActivity extends MainActivity {

    private TextView signTextView;
    private RelativeLayout usernameRelativeLayout;
    private RelativeLayout confirmPasswordRelativeLayout;
    private LinearLayout forgetPasswordLinearLayout;
    private Button signButton;
    private LinearLayout orLinearLayout;
    private TextView backToTextView;
    private Boolean viewSignIn = true;

    private EditText username, email, password, confirmPassword;

    private ProgressBar progressBar;
    String userUid;

    private UserViewModel userViewModel;
    FirebaseAuth firebaseAuth;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign);

        firebaseAuth = FirebaseAuth.getInstance();


        initialize();
        checkWhatToShowOnLayout();
    }

    private void initialize() {
        signTextView = findViewById(R.id.signTextView);
        usernameRelativeLayout = findViewById(R.id.usernameRelativeLayout);
        confirmPasswordRelativeLayout = findViewById(R.id.confirmPasswordRelativeLayout);
        forgetPasswordLinearLayout = findViewById(R.id.forgetPasswordLinearLayout);
        signButton = findViewById(R.id.signButton);
        orLinearLayout = findViewById(R.id.orLinearLayout);
        backToTextView = findViewById(R.id.backToTextView);
        username = findViewById(R.id.username);
        email = findViewById(R.id.email);
        confirmPassword = findViewById(R.id.confirmPassword);
        password = findViewById(R.id.password);
        progressBar = findViewById(R.id.progressBar_cyclic);

        if (firebaseAuth.getCurrentUser() != null) {
            startActivity(new Intent(getApplicationContext(), SecondActivity.class));
            finish();
        }
    }

    @SuppressLint("SetTextI18n")
    private void checkWhatToShowOnLayout() {
        signTextView.setOnClickListener(view -> {
            if (this.viewSignIn) {
                usernameRelativeLayout.setVisibility(View.VISIBLE);
                confirmPasswordRelativeLayout.setVisibility(View.VISIBLE);
                forgetPasswordLinearLayout.setVisibility(View.GONE);
                signButton.setText("Sign Up");
                orLinearLayout.setVisibility(View.GONE);
                backToTextView.setVisibility(View.VISIBLE);
                signTextView.setText("Sign In");
                this.viewSignIn = false;
            } else {
                usernameRelativeLayout.setVisibility(View.GONE);
                confirmPasswordRelativeLayout.setVisibility(View.GONE);
                forgetPasswordLinearLayout.setVisibility(View.VISIBLE);
                signButton.setText("Sign In");
                orLinearLayout.setVisibility(View.VISIBLE);
                backToTextView.setVisibility(View.GONE);
                signTextView.setText("Sign Up");
                this.viewSignIn = true;
            }
        });
    }

    public void showHidePass(View view) {
        if (view.getId() == R.id.hideImageView) {
            if (password.getTransformationMethod().equals(PasswordTransformationMethod.getInstance())) {
                ((ImageView) view).setImageResource(R.drawable.ic_eye);
                //Show Password
                password.setTransformationMethod(HideReturnsTransformationMethod.getInstance());
            } else {
                ((ImageView) (view)).setImageResource(R.drawable.ic_hide);
                //Hide Password
                password.setTransformationMethod(PasswordTransformationMethod.getInstance());
            }
        }
    }

    public void onSignButtonPress(View v) {
        String textSignButton = this.signButton.getText().toString();
        String email = this.email.getText().toString().trim();
        String password = this.password.getText().toString().trim();
        String username = this.username.getText().toString().trim();

        if (isFormValid()) {
            if (textSignButton.equals("Sign In")) {
                logIn(email, password);
            } else {
                signUp(email, password, username);
            }
        }
    }

    private void logIn(String email, String password) {
        if (firebaseAuth.getCurrentUser() != null)
            userUid = (firebaseAuth.getCurrentUser()).getUid();

        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if (task.isSuccessful()) {
                    DatabaseReference reference = FirebaseDatabase.getInstance().getReference().child("Users").child(firebaseAuth.getCurrentUser().getUid());
                    reference.addValueEventListener(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot snapshot) {
                            Intent intent = new Intent(SignActivity.this, SecondActivity.class);
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                            startActivity(intent);
                            finish();
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError error) {
                            progressBar.setVisibility(View.GONE);
                        }
                    });
                } else {
                    SignActivity.this.errorAtSign(task);
                }
            }
        });
    }

    private void signUp(String email, String password, String username) {
        firebaseAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                String userUid = firebaseAuth.getCurrentUser().getUid();
                User user = new User(userUid, email, "", 0, 0, "",
                        "https://firebasestorage.googleapis.com/v0/b/let-s-plant-f845c.appspot.com/o/placeholder_profileImage.png?alt=media&token=ad5ae128-f579-40e8-ad90-0fccaeda16c7",
                        username);

                userViewModel = new ViewModelProvider(this).get(UserViewModel.class);
                userViewModel.getUserMutableLiveData().observe(this, item -> {
                    userViewModel.setUserMutableLiveData(user);
                });

                databaseReference.child(DB_USERS).child(userUid).setValue(user).addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        Intent intent = new Intent(getApplicationContext(), SecondActivity.class);
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(intent);
                        finish();
                    }
                });

            } else {
                errorAtSign(task);
            }
        });
    }

    private boolean isFormValid() {
        String email = this.email.getText().toString().trim();
        String password = this.password.getText().toString().trim();
        String username = this.username.getText().toString().trim();
        String confirmPassword = this.confirmPassword.getText().toString().trim();

        if (TextUtils.isEmpty(email)) {
            this.email.setError("Email is required!");
            return false;
        }
        if (TextUtils.isEmpty(password)) {
            this.password.setError("Password is required!");
            return false;
        }
        if (password.length() < 6) {
            this.password.setError("Password must be >= 6  characters");
            return false;
        }
        if (TextUtils.isEmpty(username) && usernameRelativeLayout.getVisibility() == View.VISIBLE) {
            this.username.setError("Username is required!");
            return false;
        }
        if (username.length() < 4 && usernameRelativeLayout.getVisibility() == View.VISIBLE) {
            this.username.setError("Username must be >= 4  characters");
            return false;
        }
        if (!confirmPassword.equals(password) && confirmPasswordRelativeLayout.getVisibility() == View.VISIBLE) {
            this.confirmPassword.setError("Password does not match!");
            return false;
        }
        progressBar.setVisibility(View.VISIBLE);
        return true;
    }

    private void errorAtSign(Task<AuthResult> task) {
        Toast.makeText(SignActivity.this, Objects.requireNonNull(task.getException()).getMessage(), Toast.LENGTH_SHORT).show();
        progressBar.setVisibility(View.GONE);
    }

    public void onClickForgotPassword(View v) {
        EditText resetMail = new EditText(v.getContext());
        AlertDialog.Builder passwordResetDialog = new AlertDialog.Builder(v.getContext());
        passwordResetDialog.setTitle("Reset password?");
        passwordResetDialog.setMessage("Enter your email to receive reset link");
        passwordResetDialog.setView(resetMail);

        passwordResetDialog.setPositiveButton("Yes", (dialog, which) -> {
            String mail = resetMail.getText().toString().trim();
            if (TextUtils.isEmpty(mail)) {
                resetMail.setError("Email is required!");
                return;
            }
            firebaseAuth.sendPasswordResetEmail(mail).addOnSuccessListener(new OnSuccessListener<Void>() {
                @Override
                public void onSuccess(Void aVoid) {
                    Toast.makeText(SignActivity.this, "Reset link sent to your email", Toast.LENGTH_SHORT).show();
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    Toast.makeText(SignActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
        }).setNegativeButton("No", (dialog, which) -> {
        });
        passwordResetDialog.create().show();
    }
}

File: LetsPlant/app/src/main/java/com/e/letsplant/activities/SignActivity.kt
package com.e.letsplant.activities

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.text.method.HideReturnsTransformationMethod
import android.text.method.PasswordTransformationMethod
import android.view.*
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.ViewModelProvider
import com.e.letsplant.R
import com.e.letsplant.data.*
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import java.util.*

class SignActivity : MainActivity() {
    private var signTextView: TextView? = null
    private var usernameRelativeLayout: RelativeLayout? = null
    private var confirmPasswordRelativeLayout: RelativeLayout? = null
    private var forgetPasswordLinearLayout: LinearLayout? = null
    private var signButton: Button? = null
    private var orLinearLayout: LinearLayout? = null
    private var backToTextView: TextView? = null
    private var viewSignIn: Boolean = true
    private var username: EditText? = null
    private var email: EditText? = null
    private var password: EditText? = null
    private var confirmPassword: EditText? = null
    private var progressBar: ProgressBar? = null
    var userUid: String? = null
    private var userViewModel: UserViewModel? = null
    override var firebaseAuth: FirebaseAuth? = null

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign)
        firebaseAuth = FirebaseAuth.getInstance()
        initialize()
        checkWhatToShowOnLayout()
    }

    private fun initialize() {
        signTextView = findViewById(R.id.signTextView)
        usernameRelativeLayout = findViewById(R.id.usernameRelativeLayout)
        confirmPasswordRelativeLayout = findViewById(R.id.confirmPasswordRelativeLayout)
        forgetPasswordLinearLayout = findViewById(R.id.forgetPasswordLinearLayout)
        signButton = findViewById(R.id.signButton)
        orLinearLayout = findViewById(R.id.orLinearLayout)
        backToTextView = findViewById(R.id.backToTextView)
        username = findViewById(R.id.username)
        email = findViewById(R.id.email)
        confirmPassword = findViewById(R.id.confirmPassword)
        password = findViewById(R.id.password)
        progressBar = findViewById(R.id.progressBar_cyclic)
        if (firebaseAuth!!.currentUser != null) {
            startActivity(Intent(applicationContext, SecondActivity::class.java))
            finish()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun checkWhatToShowOnLayout() {
        signTextView!!.setOnClickListener {
            if (viewSignIn) {
                usernameRelativeLayout!!.visibility = View.VISIBLE
                confirmPasswordRelativeLayout!!.visibility = View.VISIBLE
                forgetPasswordLinearLayout!!.visibility = View.GONE
                signButton!!.text = "Sign Up"
                orLinearLayout!!.visibility = View.GONE
                backToTextView!!.visibility = View.VISIBLE
                signTextView!!.text = "Sign In"
                viewSignIn = false
            } else {
                usernameRelativeLayout!!.visibility = View.GONE
                confirmPasswordRelativeLayout!!.visibility = View.GONE
                forgetPasswordLinearLayout!!.visibility = View.VISIBLE
                signButton!!.text = "Sign In"
                orLinearLayout!!.visibility = View.VISIBLE
                backToTextView!!.visibility = View.GONE
                signTextView!!.text = "Sign Up"
                viewSignIn = true
            }
        }
    }

    fun showHidePass(view: View) {
        if (view.id == R.id.hideImageView) {
            if ((password!!.transformationMethod == PasswordTransformationMethod.getInstance())) {
                (view as ImageView).setImageResource(R.drawable.ic_eye)
                //Show Password
                password!!.transformationMethod = HideReturnsTransformationMethod.getInstance()
            } else {
                ((view) as ImageView).setImageResource(R.drawable.ic_hide)
                //Hide Password
                password!!.transformationMethod = PasswordTransformationMethod.getInstance()
            }
        }
    }

    fun onSignButtonPress() {
        val textSignButton: String = signButton!!.text.toString()
        val email: String = email!!.text.toString().trim { it <= ' ' }
        val password: String = password!!.text.toString().trim { it <= ' ' }
        val username: String = username!!.text.toString().trim { it <= ' ' }
        if (isFormValid) {
            if ((textSignButton == "Sign In")) {
                logIn(email, password)
            } else {
                signUp(email, password, username)
            }
        }
    }

    private fun logIn(email: String, password: String) {
        if (firebaseAuth!!.currentUser != null) userUid =
            (firebaseAuth!!.currentUser)!!.uid
        firebaseAuth!!.signInWithEmailAndPassword(email, password)
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    val reference: DatabaseReference =
                        FirebaseDatabase.getInstance().reference.child("Users").child(
                            firebaseAuth!!.currentUser!!.uid
                        )
                    reference.addValueEventListener(object : ValueEventListener {
                        override fun onDataChange(snapshot: DataSnapshot) {
                            val intent =
                                Intent(this@SignActivity, SecondActivity::class.java)
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                            startActivity(intent)
                            finish()
                        }

                        override fun onCancelled(error: DatabaseError) {
                            progressBar!!.visibility = View.GONE
                        }
                    })
                } else {
                    errorAtSign(task)
                }
            }
    }

    private fun signUp(email: String, password: String, username: String) {
        firebaseAuth!!.createUserWithEmailAndPassword(email, password)
            .addOnCompleteListener { task: Task<AuthResult?> ->
                if (task.isSuccessful) {
                    val userUid: String = Objects.requireNonNull(
                        firebaseAuth!!.currentUser
                    )!!.uid
                    val user = User(
                        userUid, email, "", 0.0, 0.0, "",
                        "https://firebasestorage.googleapis.com/v0/b/let-s-plant-f845c.appspot.com/o/placeholder_profileImage.png?alt=media&token=ad5ae128-f579-40e8-ad90-0fccaeda16c7",
                        username
                    )
                    userViewModel = ViewModelProvider(this).get(UserViewModel::class.java)
                    userViewModel!!.getUserMutableLiveData().observe(
                        this
                    ) { userViewModel!!.setUserMutableLiveData(user) }
                    databaseReference!!.child(DB_USERS).child(userUid)
                        .setValue(user).addOnCompleteListener {
                            val intent =
                                Intent(applicationContext, SecondActivity::class.java)
                            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                            startActivity(intent)
                            finish()
                        }
                } else {
                    errorAtSign(task)
                }
            }
    }

    private val isFormValid: Boolean
        get() {
            val email: String = email!!.text.toString().trim { it <= ' ' }
            val password: String = password!!.text.toString().trim { it <= ' ' }
            val username: String = username!!.text.toString().trim { it <= ' ' }
            val confirmPassword: String = confirmPassword!!.text.toString().trim { it <= ' ' }
            if (TextUtils.isEmpty(email)) {
                this.email!!.error = "Email is required!"
                return false
            }
            if (TextUtils.isEmpty(password)) {
                this.password!!.error = "Password is required!"
                return false
            }
            if (password.length < 6) {
                this.password!!.error = "Password must be >= 6  characters"
                return false
            }
            if (TextUtils.isEmpty(username) && usernameRelativeLayout!!.visibility == View.VISIBLE) {
                this.username!!.error = "Username is required!"
                return false
            }
            if (username.length < 4 && usernameRelativeLayout!!.visibility == View.VISIBLE) {
                this.username!!.error = "Username must be >= 4  characters"
                return false
            }
            if (confirmPassword != password && confirmPasswordRelativeLayout!!.visibility == View.VISIBLE) {
                this.confirmPassword!!.error = "Password does not match!"
                return false
            }
            progressBar!!.visibility = View.VISIBLE
            return true
        }

    private fun errorAtSign(task: Task<AuthResult?>) {
        Toast.makeText(
            this@SignActivity,
            Objects.requireNonNull(task.exception)!!.message,
            Toast.LENGTH_SHORT
        ).show()
        progressBar!!.visibility = View.GONE
    }

    fun onClickForgotPassword(v: View) {
        val resetMail = EditText(v.context)
        val passwordResetDialog: AlertDialog.Builder = AlertDialog.Builder(v.context)
        passwordResetDialog.setTitle("Reset password?")
        passwordResetDialog.setMessage("Enter your email to receive reset link")
        passwordResetDialog.setView(resetMail)
        passwordResetDialog.setPositiveButton(
            "Yes"
        ) { _: DialogInterface?, _: Int ->
            val mail: String = resetMail.text.toString().trim { it <= ' ' }
            if (TextUtils.isEmpty(mail)) {
                resetMail.error = "Email is required!"
                return@setPositiveButton
            }
            firebaseAuth!!.sendPasswordResetEmail(mail)
                .addOnSuccessListener {
                    Toast.makeText(
                        this@SignActivity,
                        "Reset link sent to your email",
                        Toast.LENGTH_SHORT
                    ).show()
                }.addOnFailureListener { e ->
                    Toast.makeText(
                        this@SignActivity,
                        e.message,
                        Toast.LENGTH_SHORT
                    ).show()
                }
        }.setNegativeButton(
            "No"
        ) { _: DialogInterface?, _: Int -> }
        passwordResetDialog.create().show()
    }
}